### PR TITLE
[pkg/ottl] Add ParseKeyValue function

### LINKED ISSUE #30998: [pkg/ottl] Add function for parsing key value pairs
### Component(s)

pkg/ottl, processor/transform

### Is your feature request related to a problem? Please describe.

I'd like to be able to parse key value pairs using OTTL similar to the [Stanza key value parser operator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/docs/operators/key_value_parser.md), but currently there is no function in OTTL that achieves this. 

### Describe the solution you'd like

Add a new converter function that returns a `pcommon.Map` of the key value pairs parsed from a target string.
```
ParseKeyValuePairs(target, Optional[delimiter], Optional[pair_delimiter])
```
The function parameters would be as follows:
- `target`: `StringGetter` Returns a string containing key value pairs to be parsed from such as `"pkg=ottl func=keyvalue"`
- `delimiter`: `Optional[string]` A string containing the delimiter value to use to split the key value pair, such as `"="`. Default can be `=`.
- `pair_delimiter`: `Optional[string]`A string containing the delimiter value to use to split pairs in the target string. Default can be white space.

Implementation can follow the Stanza package key value parser operator and use `strings.FieldsFunc()` when white space is the `pair_delimiter`, `strings.Split()` for non white space `pair_delimiters`, and `strings.SplitN()` for splitting the key and value in a given pair. 

### Describe alternatives you've considered

A possible alternative is implementing this function as an editor function. This option would allow users to immediately overwrite the source of the key value pairs with the newly parsed version. However, there is a precedent already set with the `ParseJSON` converter function to just return the parsed value. While it may be useful for this function to also replace the source string, I don't think it's necessary for it to do so and just returning a map of the pairs is enough.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dpaasman00**
**Description:** <Describe what has changed.>
Adds a `ParseKeyValue` converter function that parses out key values pairs into a `pcommon.Map`. It takes a `StringGetter` target argument and 2 optional arguments for the pair delimiter and key value delimiter. This is an adaptation of the Stanza Key Value Parser operator to provide feature parity. 

Given the following input string `"k1=v1 k2=v2 k3=v3"`, the function would return the following map:
```
{ "k1": "v1", "k2": "v2", "k3": "v3" }
```

**Link to tracking Issue:** <Issue number if applicable>
Closes #30998 

**Testing:** <Describe what testing was performed and which tests were added.>
Added unit tests and e2e test.

**Documentation:** <Describe the documentation added.>
Added function documentation.

### CODE REVIEW

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (BinaryFissionGames) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:43]: Should this check be after the empty check? As it's written, I could set delimiter to `" "` and have pair_delimiter == delimiter.

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:43]: Yes, good catch, thank you!

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (BinaryFissionGames) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:90]: I think this is the same as the Stanza version of the parser, but it looks like you can start/end a quoted section with different quote types?

e.g. (where pair_delimiter is `" "`, delimiter is `"="`) 
This will parse: `key='this is a value"`
This will not parse: `key='this is a "cool" value'`

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dehaansa) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:73]: Could use a comment on why we're not just using strings.Split (because of quoted values)

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dehaansa) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:79]: Readability wise, this loop is kind of hard. For example `i += delimiterLength - 1` is confusing when delimiterLength == 1 (normal case), as it's effectively a noop.

I would favor refactoring this to a "while" equivalent loop. Something like the following (untested)
```
i := 0
for i < len(input) {
	if !inQuotes && i+delimiterLength <= len(input) && input[i:i+delimiterLength] == delimiter {
		if currentPair == "" {
			continue
		}
		result = append(result, currentPair)
		currentPair = ""
		i += delimiterLength
		continue
	}
	
	if input[i] == `"` || input[i] == `'` {
		inQuotes = !inQuotes
	}

	currentPair += string(input[i])
	i++
}
```

File: pkg/ottl/ottlfuncs/func_parse_key_value_test.go
Comment (dehaansa) [inline pkg/ottl/ottlfuncs/func_parse_key_value_test.go:61]: Is there a test case with the pair_delimiter present in a quoted value? I didn't see one but I may have missed it.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:90]: Ah yea that's a good catch. I can confirm the first example will be parsed successfully even though it shouldn't be. The second example as is does pass but only because the pair delimiter isn't present. The following string is what will fail due to the mismatching quotes situation. 
- `key='this is a "co ol" value'`

Regardless, I'll change it to keep track of which type of quote is being used. Again, nice catch, thank you!

File: pkg/ottl/ottlfuncs/func_parse_key_value_test.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value_test.go:61]: Yep, the test `quoted value contains delimiter and pair delimiter` covers that case. Although if you think it'd be better to split those 2 cases up/each case has their own specific test case I can certainly do so

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:79]: Yea I agree, the logic being used is kind of messy/hard to follow so I'm in favor of simplifying where it makes sense. I like the idea of changing to the while loop at least to get rid of that "sneaky" noop.

File: pkg/ottl/ottlfuncs/func_parse_key_value_test.go
Comment (dehaansa) [reply pkg/ottl/ottlfuncs/func_parse_key_value_test.go:61]: Nope, that's fine. I was just blind 🙈

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:115]: not sure how I feel about this error message, thoughts?

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (BinaryFissionGames) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:38]: nit: could we have `pair_delimiter` be mixedCaps instead of snake_case?
https://go.dev/doc/effective_go#mixed-caps

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (BinaryFissionGames) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:81]: Could we rename delimiter here to `pairDelimiter`? Just makes it more obvious which delimiter is being used here.

Also, thoughts on renaming this function into `splitIntoPairs` or maybe just `splitPairs`, since it's a bit more specific than just `splitString`?

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dehaansa) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:83]: Do we need inQuote && quoteType? Couldn't inQuote just be `quoteType != ""`?

**Review by dehaansa — APPROVED** [review]

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (djaglowski) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:45]: Should this be set and validated in `createParseKeyValueFunction`? I'm not very familiar with OTTL code patterns but generally in the collector we want configuration errors to be detected when instantiating components, rather than when processing data.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (djaglowski) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:81]: Since this is pretty complex and already implemented in `pkg/stanza`, we should consider moving it to a shared internal package. To my knowledge, there's not currently a lot of shared code between ottl and pkg/stanza but we have done this at least with `internal/coreinternal/timeutils`. So perhaps a new `internal/coreinternal/parseutils` package would be appropriate. I'm curious what @TylerHelmuth and/or @evan-bradley think about this.

**Review by djaglowski — COMMENTED** [review]
I'm not familiar enough with ottl code standards to approve this but I can confirm it appears to implement equivalent parsing functionality to `keyvalue_parser` from `pkg/stanza`

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:45]: I'm not entirely sure what the answer is here. I was able to find a few other examples of configuration error handling in the data processing function elsewhere in OTTL such as the `limit` function [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/func_limit.go#L36-L38). Maybe one of the maintainers can give more insight? @TylerHelmuth @evan-bradley

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:84]: I agree, even if just this parser ends up in there, I think shared internal packages have low enough overhead to make one worth it here.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:69]:
```suggestion
				return nil, fmt.Errorf("cannot split %q into 2 items, got %d items", p, len(pair))
```
Same here for the format specifier, and just a minor word nit.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: Can you add tests for what happens if either of these are an empty string? Also what happens if they are longer than the target, or aren't found in the target?

**Review by evan-bradley — COMMENTED** [review]
Looks good overall, just a few minor notes.

File: pkg/ottl/ottlfuncs/README.md
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/README.md:850]: Nice callout

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: Good catch, the current implementation allows either to be equal to `""` which should not be allowed. In the case one of the delimiters is set to an empty string, should the function error or use the default?

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: I could easily be misunderstanding something, but trying the following test case (`expected` doesn't matter here):
```golang
               {
			name: "empty delimiters",
			target: ottl.StandardStringGetter[any]{
				Getter: func(ctx context.Context, tCtx any) (any, error) {
					return "a=b c=d", nil
				},
			},
			delimiter:      ottl.NewTestingOptional(""),
			pair_delimiter: ottl.NewTestingOptional(""),
			expected:       map[string]any{},
		},
```
Errors on the check for equal delimiters in the start of `parseKeyValue()`:
`Error: Received unexpected error: pair delimiter "" cannot be equal to delimiter ""`

Which tells me it is possible for the delimiter variables to get set with the empty string if the user passes them in. Obviously in this case they can't both be set because it gets caught, but it demonstrates that one of them can be set to `""`.

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: Unless you're referring to when the OTTL function would be used while the collector is running and not isolated like in a unit test?

Comment (djaglowski) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: I'm referring to what the user may specify in their config. Whether they omit either parameter or explicitly set it to `""`, it should change the delimiter to the default value.

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:18]: Ahhh, yep, I misunderstood your first comment. That behavior works for me and I can change it to that.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:48]: I'll leave this up to @TylerHelmuth

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:19]: >Whether they omit either parameter or explicitly set it to `""`, it should change the delimiter to the default value.

We can check if the user passed a parameter using the `IsEmpty()` function on the `Optional` type. In my opinion, if we want to forbid these values, we should check if they have been explicitly set to an invalid value and throw an error instead of using the default, or at a minimum print a warning. I would be surprised if the value I set was silently ignored by the function.

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:19]: I'll have it throw an error since I don't really see an example of logging a warning in another OTTL function.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:68]: Can we emit a debug/warning log if the pairDelimiter isn't found? I think this is something we would want to alert the operator to while still trying to emit what data we can. You can get a logger from `ottl.FunctionContext`.

@TylerHelmuth What are your thoughts on this? I think it would make sense to emit what data we can, but maybe we should just return an error here. I don't think there's much precedence for this in OTTL, usually we just return an error.

File: pkg/ottl/ottlfuncs/func_parse_key_value_test.go
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/func_parse_key_value_test.go:351]: Could we have a similar tets for `a==b c=d` or something similar that would produce too many pairs?

**Review by evan-bradley — COMMENTED** [review]
Thanks for your persistence, I think we're getting close.

File: pkg/ottl/ottlfuncs/func_parse_key_value_test.go
Comment (evan-bradley) [reply pkg/ottl/ottlfuncs/func_parse_key_value_test.go:351]: Got it, that makes sense. I think we are good here then, thank you.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:65]: Yea that's a good point. It can definitely be changed to just a trivial case, but I do think there's value in getting some kind of feedback about the target if it is empty to aid in configuring the function. I'm not sure though, maybe that's not really a big point of concern, in which case making it trivial seems fine to me.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:68]: A debug may be best because in the case of malformed input the results of further parsing will be totally broken, and it may not be totally obvious why. It could be as simple as, "no pair delimiter detected, possible malformed input" or something of the like.

Comment (djaglowski) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:68]: Special cases aside, I think it's very common that users will just want the default values for both delimiter and pair delimiter. I wouldn't emit warnings if using the defaults.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go]: Sharing via `internal` sounds good to me.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go]: Validating parameters is typically done in the function before the `ExprFunc` is returned.  I'm good with keeping validation where is currently is.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:65]: [stanza throws an error so we should too.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6f1b1d2a5c2a690e138feac4712c7af01b65f2a4/pkg/stanza/operator/parser/keyvalue/keyvalue.go#L97-L99)

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:68]: @evan-bradley I'm not really clear on the concern.  I don't see stanza handling that situation as a special case so I don't think we need to either.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_parse_key_value.go:82]: I think this bit of code is the same as https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6f1b1d2a5c2a690e138feac4712c7af01b65f2a4/pkg/stanza/operator/parser/keyvalue/keyvalue.go#L106-L120, can it also be moved to `coreinternal`?

Comment (dpaasman00) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:82]: Yea definitely, only question I have is how do we want to handle the error that occurs when the delimiter isn't found in the pair string? In Stanza it parses all pairs before returning where as the current OTTL implementation returns as soon as an error is encountered.

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:82]: I think following stanza here is ok.  We'll return a detailed error, which is helpful.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:65]: Thanks for calling that out. I just wanted to explore that option, but I think we're good here.

File: pkg/ottl/ottlfuncs/func_parse_key_value.go
Comment (evan-bradley) [reply pkg/ottl/ottlfuncs/func_parse_key_value.go:82]: I'm fine just matching Stanza's behavior. If in the future we want to change it in one place it probably makes sense to change it in both.

**Comment by dpaasman00** [comment]
Unsure if I should update Stanza in this PR similar to how it was done in #31081 or if that's a separate PR?

**Review by TylerHelmuth — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**
